### ISSUE TITLE
Determine user eligibility

### ISSUE THREAD

**Issue by biancadanforth** [biancadanforth/gecko-dev#13]
To what extent can/should the user's eligibility be determined by Normandy versus the core study code?

I know we can target locale and geolocation via Normandy, so that covers that aspect. How do we identify "new" and "existing" users as defined in the study PHD?:
* new users
  * > profile was created in the past minutes
* existing users
  * > profiles that have been active at least once every 2 weeks for the past 3 months...that were created at least 3 weeks ago.

I should be able to identify profile age in the code:
Per Mossop, in the user profile, there’s a `times.json` that has a “created” key. I can use the `ProfileAge.jsm` module and call `const creationDate = ProfileAge.created;`.

I am less certain about how to measure activity (i.e. they've been "active at least once every 2 weeks for the past 3 months"). @gregglind, you mentioned this might require digging through all the pings a profile has sent, can you elaborate on how I would pull the pings and what I would look for?

**Comment by biancadanforth** [comment]
After discussing this with @SuYoungHong, Normandy/Bianca will handle new versus existing users via PCD (profile creation date) with different experiment slugs, see #9 , and existing users will be filtered out by a TBD definition of "active" by Su during the analysis.

I need to know from the stakeholders how old a profile can be to be considered an "existing" user. The PHD states both 3 months and 3 weeks. I have added a [comment in the PHD](https://docs.google.com/a/mozilla.com/document/d/1emT5bdFQoMh5Zl0bLZS5fHKWKhXRpLrptDesKdfQYCw/edit?disco=AAAABxdUNvs) to clarify this.

Note: There is an open investigation into PCD that Su is working on. See [this](https://docs.google.com/presentation/d/19cToJL2JY4YnLOmnrAPe_EQ_VGSjjANXo-3yMy-9oKk/edit#slide=id.p).

**Comment by SuYoungHong** [comment]
+1

**Comment by biancadanforth** [comment]
I just clarified with chutten/mythmon about how to target **brand** new users in the first minutes of their first session:

### TL;DR:
The client for a brand new profile will ping Normandy on the first start-up (this usually happens within seconds of the first session). In this first instance, `normandy.isFirstRun` will be true. To keep this brand new user enrolled for the duration of the study (so that the filter expression still evaluates to `true` after their first session), we will also use `profileCreationDate`; combining these, the expression will look something like: `isFirstRun || PCD > magicNumber`. Note that PCD is only accurate to the DAY.
On the telemetry side of things, as soon as Normandy flips the pref (within seconds of a new users first Firefox startup), assuming my study code correctly enables my study category (`"savant"`) of events, event telemetry for that category will start being appended to the main ping. No telemetry is sent from the client's machine in the first 30 minutes of a new user's first session as a rule (or during their first session if its duration is less than 30 minutes) for privacy reasons; however, that main ping will be sent on first-shutdown (or, if the session is longer than 30 minutes, possibly when a new sub-session is triggered, see [main ping docs](https://firefox-source-docs.mozilla.org/toolkit/components/telemetry/telemetry/data/main-ping.html)).

**My conversation with chutten on using telemetry data for targeting**

> there are known issues with PCD that we are investigating, so it shouldn't be used in isolation. The "new-profile" ping is a good source of information for when users are starting a new profile

What are the known issues with PCD?

> So, PCD is unreliable because it sometimes relies on the age of files in the profile directory to determine the age of a profile and that... well, it isn't a great approach: https://mzl.la/2He4Lj1 shows that 10% of files in profile directories scanned for age were > 200 days old

But looking at the ["new-profile" ping](https://firefox-source-docs.mozilla.org/toolkit/components/telemetry/telemetry/data/new-profile-ping.html), that isn't sent until 30 minutes into the first session or on first session shutdown (whichever occurs first). So we couldn't use this either. What should I do?

> there is never going to be data sent within 30min of a new profile
> That's the length of time we provide for a user to familiarize themselves with our Privacy Policy and decide whether they want to trust us, or if they want to disable Telemetry
> We won't send data until the user has a chance to decide not to. This is a legal requirement (and good sense) so it is unlikely to change
> There is one time that we'll send early: if the user decides to terminate that first session earlier than 30min and without disabling Telemetry.
> This happens about 78% of the time on beta
> (or did when we started sending new-profile pings to measure it: https://reports.telemetry.mozilla.org/post/projects%2Fnewprofile_ping_beta_validation.kp )

So any event telemetry from the first session will be sent in that shutdown main ping right?

 > In the "first-shutdown", correct
> It may change soonish that we'll be sending it in an "event" ping instead of "first-shutdown", but that first session has coverage (unless the user turns us off)

**My conversation with mythmon in using available properties in Normandy (which includes access to telemetry data) for targeting**

if I wanted to target a new user in the first minutes of their first session, what would be the recommended way to do that with a Normandy filter expression?
PCD is only accurate to the day I believe
the "new-profile" ping is only after 30 minutes or when the first session ends

> so `isFirstRun` would evaluate to true the first time Normandy run, and then false the next time
so the pref would flip, and then unflip 6 hours later
> the recommended approach is to choose a point in time, and then take any profile with PCD > that day
> but since PCD isn't available at first run, you end up doing something like `isFirstRun || PCD > magicNumber`

What happens every 6 hours? I don't follow why it unflips at that point

> normandy runs every 6 hours (now)

What do you mean by run?

> normandy runs on a timer. every time the timer fires, Normandy wakes up, downloads a list of recipes from the server, evaluates their filters, executes the one that matches, and cleans up any that stopped matching

But a new profile pings Normandy to run right away right? Why would we also want PCD?

> because the second time Normandy runs, `isFirstRun` will no longer be true, and so the filter will no longer match the user. that will cause the user to unenroll in the study

 so Normandy runs every 6 hours, but also the client for a new profile will ping it to run as well right away right?

> due to a silly bug, it will run twice at startup. once immediately, and then once when the timer notices that it has been more than 6 hours since the timer last fired (forever is more than 6 hours)
and by "at startup" I mean at first startup of a new profile

**Comment by biancadanforth** [comment]
@MattGrimes, @rayborn, here are some notes I took on deploying the study; might be helpful for launching next week.

**Comment by biancadanforth** [comment]
Closing, since we have no more questions around this, and the study has launched.
